cipher/ascon/ascon.go

```diff
@@ -0,0 +1,260 @@
+// and Open.
+func (a *Cipher) NonceSize() int { return NonceSize }
+
+// Overhead returns the maximum difference between the lengths of a
+// plaintext and its ciphertext.
+func (a *Cipher) Overhead() int { return TagSize }
+
+// Seal encrypts and authenticates plaintext, authenticates the
+// additional data and appends the result to dst, returning the updated
+// slice. The nonce must be NonceSize() bytes long and unique for all
+// time, for a given key.
+//
+// To reuse plaintext's storage for the encrypted output, use plaintext[:0]
+// as dst. Otherwise, the remaining capacity of dst must not overlap plaintext.
+func (a *Cipher) Seal(dst, nonce, plaintext, additionalData []byte) []byte {
+	if len(nonce) != NonceSize {
+		panic(ErrNonceSize)
+	}
+
+	ptLen := len(plaintext)
+	output := make([]byte, ptLen+TagSize)
+	ciphertext, tag := output[:ptLen], output[ptLen:]
+
+	a.initialize(nonce)
+	a.assocData(additionalData)
+	a.procText(plaintext, ciphertext, true)
+	a.finalize(tag)
+
+	return output
+}
+
+// Open decrypts and authenticates ciphertext, authenticates the
+// additional data and, if successful, appends the resulting plaintext
+// to dst, returning the updated slice. The nonce must be NonceSize()
+// bytes long and both it and the additional data must match the
+// value passed to Seal.
+//
+// To reuse ciphertext's storage for the decrypted output, use ciphertext[:0]
+// as dst. Otherwise, the remaining capacity of dst must not overlap plaintext.
+//
+// Even if the function fails, the contents of dst, up to its capacity,
+// may be overwritten.
+func (a *Cipher) Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error) {
+	if len(nonce) != NonceSize {
+		panic(ErrNonceSize)
+	}
+
+	ptLen := len(ciphertext) - TagSize
+	plaintext := make([]byte, ptLen)
+	ciphertext, tag0 := ciphertext[:ptLen], ciphertext[ptLen:]
+	tag1 := (&[TagSize]byte{})[:]
+
+	a.initialize(nonce)
+	a.assocData(additionalData)
+	a.procText(ciphertext, plaintext, false)
+	a.finalize(tag1)
+
+	if !bytes.Equal(tag0, tag1) {
+		return nil, ErrDecryption
+	}
+
+	return plaintext, nil
+}
+
+func (a *Cipher) initialize(nonce []byte) {
+	bcs := blockSize * byte(a.mode)
+	pB := permB + 2*(byte(a.mode)-1)
+	a.state[0] = KeySize * 8
+	a.state[1] = bcs * 8
+	a.state[2] = permA
+	a.state[3] = pB
+	a.state[4] = 0
+	a.state[5] = 0
+	a.state[6] = 0
+	a.state[7] = 0
+	copy(a.state[ivSize:ivSize+KeySize], a.key[:])
+	copy(a.state[ivSize+KeySize:ivSize+KeySize+NonceSize], nonce)
+	a.perm(permA)
+
+	for i := 0; i < KeySize; i++ {
+		a.state[stateSize-KeySize+i] ^= a.key[i]
+	}
+}
+
+func (a *Cipher) assocData(add []byte) {
+	bcs := blockSize * int(a.mode)
+	pB := permB + 2*(int(a.mode)-1)
+	if len(add) > 0 {
+		for ; len(add) >= bcs; add = add[bcs:] {
+			for i := 0; i < bcs; i++ {
+				a.state[i] ^= add[i]
+			}
+			a.perm(pB)
+		}
+		if len(add) >= 0 {
+			for i := 0; i < len(add); i++ {
+				a.state[i] ^= add[i]
+			}
+			a.state[len(add)] ^= 0x80
```

Comment:
If `add` is a multiple of `bcs` then we don't XOR in `0x80`. Is that correct?

Reply:
not sure,

Reply:
If I read §2.4.2 of the Ascon spec, then the padding can only be omitted in case the associated data is empty.

Reply:
there is an if at the top that jumps when add is empty